File: 2016/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import hashlib
from typing import Any, Generator


def solution(content: str) -> Any:
    return "".join(p[5] for p in take(8, hash_gen(content)))


def hash_gen(prefix: str):
    counter = 0
    while True:
        hashed = hashlib.md5(bytes(prefix + str(counter), "utf8")).hexdigest()
        if hashed.startswith("0" * 5):
            print(counter, hashed)
            yield hashed
        counter += 1


def take(n: int, g: Generator):
    for _ in range(n):
        yield next(g)


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2016/day05-1.ts
import { encodeHex } from "jsr:@std/encoding/hex";
import { md5 } from "jsr:@takker/md5";

function solution(content: string) {
  const g = generate(content);
  return Array.from({ length: 8 })
    .map((_) => g.next().value[5])
    .join("");
}

function* generate(prefix: string): Generator<string, string, unknown> {
  let counter = 0;
  while (true) {
    const hashed = encodeHex(md5(prefix + counter));
    if (hashed.startsWith("00000")) {
      yield hashed;
    }
    counter += 1;
  }
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
